subprojects/core/src/main/java/org/gradle/api/internal/file/copy/CopySpecActionImpl.java

```diff
@@ -15,25 +15,78 @@
  */
 package org.gradle.api.internal.file.copy;
 
+import com.google.common.base.Function;
+import com.google.common.collect.Ordering;
 import org.gradle.api.Action;
 import org.gradle.api.file.FileTree;
 import org.gradle.api.internal.file.CopyActionProcessingStreamAction;
 import org.gradle.internal.nativeintegration.filesystem.FileSystem;
 import org.gradle.internal.reflect.Instantiator;
 
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
 public class CopySpecActionImpl implements Action<CopySpecResolver> {
     private final CopyActionProcessingStreamAction action;
     private final Instantiator instantiator;
     private final FileSystem fileSystem;
+    private final boolean reproducibleFileOrder;
 
-    public CopySpecActionImpl(CopyActionProcessingStreamAction action, Instantiator instantiator, FileSystem fileSystem) {
+    public CopySpecActionImpl(CopyActionProcessingStreamAction action, Instantiator instantiator, FileSystem fileSystem, boolean reproducibleFileOrder) {
         this.action = action;
         this.instantiator = instantiator;
         this.fileSystem = fileSystem;
+        this.reproducibleFileOrder = reproducibleFileOrder;
     }
 
     public void execute(final CopySpecResolver specResolver) {
+        if (reproducibleFileOrder) {
+            SortingStreamAction sortingStreamAction = new SortingStreamAction(action);
```

Comment:
Hrm... I think this is close to where I thought it would be, but I thought it would be wrapped around the `ZipCopyAction` or `TarCopyAction`.  Something like in `CopyActionExecuter`, it would be another decorator after `NormalizingCopyActionDecorator`.  Isn't the sorting here happening before the `eachFile` hooks and the duplication handling? 

Instead of mixing this into `CopySpecActionImpl`, we should try to apply it like a decorator